File: DATA_STRUCTURES_CPP/Queue_Priority__(based_list).cpp
/*Priority Queue FIFO -- Очередь с приоритетом на основе связного списка*/
#include <iostream>

using std::cout;


template <typename Type>
class Pqueue {								// Класс описывающий очередь с приоритетом
private:
	class Node {							// Класс описывающий узел очереди
	public:
		Type data;							// Поле данных
		Node* next;							// Указатель на следующий узел списка
		int priority;						// Приоритет элемента
		Node(const Type& d, const int& p) : // Конструктор
			data(d), priority(p), next(nullptr) {};
	};
	Node* head;								// Начало очереди
	Node* tail;								// Конец очереди
	int count;								// Кол-во элементов
public:
	Pqueue();								// Конструктор
	Pqueue(const Type &elm, const int& pri);// Конструктор с параметром
	Pqueue(const Pqueue<Type> &obj);		// Конструктор копирования
	~Pqueue();								// Деструктор
	void show();							// Вывод очереди
	void pop();								// Удалить верхний (первый) элемент
	void clear();							// Очистка очереди
	const Type& front() const;				// Вернуть 1-ый элемент
	const Type& back()  const;				// Вернуть последний элемент
	int size() { return count; }			// Вернуть размер очереди
	bool empty() { return count == 0; }		// Вернуть 0 если очередь пустая, иначе 1
	void push(const Type& elm, const int& pri);		// Добавление в очередь элемента по приоритету

};

template <typename Type>
Pqueue<Type>::Pqueue() {
	head = tail = nullptr;
	count = 0;
}

template <typename Type>
Pqueue<Type>::Pqueue(const Type& elm, const int& pri) {
	Node *new_node = new Node(elm, pri);
	head = tail = new_node;
	count = 1;
}

template <typename Type>
Pqueue<Type>::Pqueue(const Pqueue<Type>& obj) {
	Node *tmp = obj.head;
	while (tmp) {
		push(tmp->data, tmp->priority);
		tmp = tmp->next;
	}
	head = obj.head;
	tail = obj.tail;
}

template <typename Type>
Pqueue<Type>::~Pqueue() {
	clear();
}

template <typename Type>
void Pqueue<Type>::push(const Type &elm, const int &prior) {
	Node* new_node = new Node(elm, prior);  // Создать новый элемент очереди
	if (head == nullptr) {					// Если очередь пустая, тогда элемент будет 1-ым
		head = new_node;					// Голова указывает на новый добавленный элемент
		tail = new_node;					// Хвост указывает на новый добавленный элемент
		count++;							// Увеличить счетчик
		return;								// Вернуться
	}
	Node* temp = head;						// Временный указатель для обхода и поиска приоритета в цикле
	Node* prev = nullptr;					// Указатель на предыдущий элеменет от temp
	// Итерироваться по очереди, пока не закончится или пока не встретится меньший приоритет
	while (temp != nullptr && temp->priority > prior) {
		prev = temp;						// В предыдущий запомнить значение текущего
		temp = temp->next;					// Текущий продвинуть вперед по очереди
	}
	if (temp == nullptr) {					// Если дошли до конца очереди и не встретился приоритет ниже чем у добавляемого
		prev->next = new_node;				// Добавить элемент в конец очереди
		tail = new_node;					// Хвост указывает на последний элемент
	}
	else {
		if (prev == nullptr) {				// Если предыдущий указвает на 0, значит итераций не было т.к приоритет выше
			new_node->next = head;			// Новый элемент встает вперед очереди
			head = new_node;				// Голова указывает на новый элемент
		}
		else {								// Если итерации были, тогда настроить указатели
			new_node->next = temp;			// Новый элемент будет указывать на найденное значение
			prev->next = new_node;			// А предыдущий будет указывать на новый
		}
	}
	count++;
}

template <typename Type>
void Pqueue<Type>::show() {
	if (count == 0) { std::cout << "Queue is empty\n"; }
	Node* tmp = head;
	for (int i = 0; i < count; ++i) {
		cout << "(" << tmp->data << ":" << tmp->priority << ")<-";
		tmp = tmp->next;
	}
}

template <typename Type>
const Type& Pqueue<Type>::front() const {
	return (count == 0) ? Type(0) : head->data;
}

template <typename Type>
const Type& Pqueue<Type>::back() const {
	return (count == 0) ? Type(0) : tail->data;
}

template <typename Type>
void Pqueue<Type>::pop() {
	if (head == nullptr) { return; }
	Node* tmp = head;
	head = head->next;
	delete tmp;
	count--;
}

template <typename Type>
void Pqueue<Type>::clear() {
	while (count) {
		pop();
	}
}


int main() {
	setlocale(0, "Ru");
	srand(static_cast<unsigned int>(time(NULL)));
	clock_t time = clock();
	Pqueue<int> pqueue;
	std::cout << "=====Начало теста=====\n";
	for (int i = 0; i < 10'000; ++i) {
		pqueue.push(rand() % 100'000, rand() % 10);
	}
	time = clock() - time;
	std::cout << "Добавление элементов: " << (double)time / CLOCKS_PER_SEC << " сек." << std::endl;
	time = clock();
	pqueue.clear();
	time = clock() - time;
	std::cout << "Извлечение элементов: " << (double)time / CLOCKS_PER_SEC << " сек." << std::endl;

	for (int i = 0; i < 15; i++)
		pqueue.push(i, i % 10);
	pqueue.show();
	return 0;
}

File: DATA_STRUCTURES_PY/Queue_Priority__(based_list).py
# Priority Queue FIFO -- Очередь с приоритетом на основе связного списка
import random

class Node: pass                                # Класс описывающий узел очереди
class Pqueue:                                   # Класс описывающий очередь
    def __init__(self, data, prior): pass       # Конструктор с параметром и без
    def push(self, elem, prior): pass           # Добавление элементов по приоритету
    def front(self): pass                       # Возвращает первый элемент в очереди, если он есть
    def back(self): pass                        # Возвращает последний элемент
    def pop(self): pass                         # Удаление из начала очереди FIFO
    def clear(self): pass                       # Очистка очереди
    def show(self): pass                        # Печать
    def empty(self): pass                       # true Если пуста, иначе false
    def size(self): pass                        # Возвращает кол-во элементов
"""----------------------------------------------------------------"""

class Node:
    def __init__(self, data, prior):
        self._data = data
        self._prior = prior
        self._next = None

class Pqueue:
    def __init__(self, data=None, prior=None):  # Конструктор с параметром и без
        if data is None and prior is None:      # Если не переданный ни элемент ни приоритет
            self._head = None                   # Инициализация полей null-ами
            self._tail = None
            self._count = 0
        else:                                   # Если элемент и приоритет переданы
            self._count = 1                     # Увеличить счетчик на 1
            node = Node(data, prior)            # Создать новый узел
            self._head = self._tail = node      # И начало и конец очереди указывают на него

    def push(self, elem, prior):                # Добавление элементов по приоритету
        new_node = Node(elem, prior)            # Создать элемент с заданным приоритетом
        if self._head is None:                  # Если это первый добавляемый элемент, то
            self._head = self._tail = new_node  # Тогда и начало и конец указывают на него
            self._count += 1
            return
        temp = self._head                       # Указатель временный на начало очереди
        prev = None                             # Запоминатель адреса
        # Пока есть элементы в очереди и приоритет выше чем у заданного, продвигаться к концу
        while (temp != None) and (temp._prior > prior):
            prev = temp                         # Запомнить адрес предыдущего элемента
            temp = temp._next                   # Продвинуться по очереди вперед
        if temp == None:                        # Если дошли до конца очереди и заданный приоритет был самым низким
            prev._next = new_node               # Добавить это элемент в конец очереди
            self._tail = new_node               # Хвост указывает на него
        else:                                   # Если встретил низший приоритет, значит дальше приоритеты ниже
            if prev == None:                    # Если предыдущий указвает на 0, значит итераций не было т.к приоритет выше
                new_node._next = self._head     # Новый элемент встает вперед очереди
                self._head = new_node           # Голова указывает на новый элемент
            else:                               # Если итерации были, тогда настроить указатели
                new_node._next = temp           # Новый элемент будет указывать на найденное значение
                prev._next = new_node           # А предыдущий будет указывать на новый
        self._count += 1                        # Увеличить счетчик

    def front(self):
        return self._head._data if self._count > 0 else None

    def back(self):
        return self._tail._data if self._count > 0 else None

    def pop(self):                              # Удаление из начала очереди FIFO
        if self._head is None:
            return
        tmp = self._head
        self._head = self._head._next
        del tmp
        self._count -= 1

    def clear(self):
        while self._count:
            self.pop()

    def show(self):
        if self._count == 0:
            print("Queue is empty!")
            return
        tmp = self._head
        while tmp:
            print(f"({tmp._data}:{tmp._prior})", end='<-')
            tmp = tmp._next
        print('')

    def empty(self):
        return self._count == 0

    def size(self):
        return self._count


def main():
    pqu = Pqueue(4, 1)
    for i in range(15):
        pqu.push(i, random.randint(0, 10))
    pqu.show()
    for i in range(5):
        pqu.pop()
    pqu.show()
    print(f"front = {pqu.front()}")
    print(f"front = {pqu.back()}")
    pqu.clear()
    pqu.show()


if __name__ == "__main__":
    main()